File: mcp-code-intelligence-python/src/mcp_code_intel/http/ingest_routes.py
"""HTTP ingest-file route — POST /api/memory/ingest-file.

Allows extension to directly index documents without going through MCP stdio.
Supports single file and batch requests with deduplication via checksum.
"""

import hashlib
import json
import os
import sys
from http.server import BaseHTTPRequestHandler
from pathlib import Path
from typing import Any


def handle_ingest_file_route(
    handler: BaseHTTPRequestHandler,
    memory_engine: Any,
    workspace: str,
) -> None:
    """Handle POST /api/memory/ingest-file."""
    if memory_engine is None:
        _send_json(handler, 503, {"error": "Memory not initialized"})
        return

    try:
        length = int(handler.headers.get("Content-Length", 0))
        body = handler.rfile.read(length).decode("utf-8")
        data = json.loads(body)
    except (json.JSONDecodeError, ValueError) as e:
        _send_json(handler, 400, {"error": f"Invalid JSON: {e}"})
        return

    files = data.get("files") or [data]
    results = [_ingest_single(f, memory_engine, workspace) for f in files]
    ingested = sum(1 for r in results if not r["skipped"])
    skipped = sum(1 for r in results if r["skipped"])

    _send_json(handler, 200, {
        "total": len(results),
        "ingested": ingested,
        "skipped": skipped,
        "results": results,
    })


def _ingest_single(req: dict, engine: Any, workspace: str) -> dict:
    """Ingest a single file with dedup check."""
    file_path = req.get("file_path", "")
    doc_type = req.get("type", "CONTEXT")

    resolved = _resolve_path(file_path, workspace)
    if not resolved or not os.path.isfile(resolved):
        return {"file_path": file_path, "entries_created": 0, "skipped": True, "reason": "file not found"}

    content = Path(resolved).read_text(encoding="utf-8")
    checksum = hashlib.md5(content.encode()).hexdigest()

    # Dedup: check if already ingested with same checksum
    if hasattr(engine, "get_source_checksum"):
        existing = engine.get_source_checksum(file_path)
        if existing == checksum:
            return {"file_path": file_path, "entries_created": 0, "skipped": True, "reason": "unchanged (checksum match)"}

    # Ingest via pipeline
    from ..memory.ingest_pipeline import ingest_markdown
    entries = ingest_markdown(engine, content, file_path, doc_type)

    if hasattr(engine, "set_source_checksum"):
        engine.set_source_checksum(file_path, checksum)

    return {"file_path": file_path, "entries_created": entries}


def _resolve_path(file_path: str, workspace: str) -> str | None:
    """Resolve file path relative to workspace."""
    if os.path.isabs(file_path) and os.path.isfile(file_path):
        return file_path
    if workspace:
        ws_path = os.path.join(workspace, file_path)
        if os.path.isfile(ws_path):
            return ws_path
    return None


def _send_json(handler: BaseHTTPRequestHandler, status: int, data: dict) -> None:
    """Send JSON response."""
    body = json.dumps(data).encode("utf-8")
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json")
    handler.send_header("Access-Control-Allow-Origin", "*")
    handler.send_header("Content-Length", str(len(body)))
    handler.end_headers()
    handler.wfile.write(body)

File: backend/mcp-code-intelligence-nodejs/src/http/ingest-routes.ts
/**
 * HTTP ingest-file route — POST /api/memory/ingest-file.
 * Allows extension to directly index documents without going through MCP stdio.
 */

import * as http from 'http';
import * as fs from 'fs';
import * as path from 'path';
import { MemoryEngine } from '../memory/memory-engine.js';

interface IngestFileRequest {
  file_path: string;
  type?: string;
  format?: string;
}

interface IngestFileResult {
  file_path: string;
  entries_created: number;
  skipped: boolean;
  reason?: string;
}

/** Handle POST /api/memory/ingest-file. */
export function handleIngestFileRoute(
  req: http.IncomingMessage,
  url: URL,
  res: http.ServerResponse,
  engine: MemoryEngine | null,
  workspace: string
): void {
  if (url.pathname !== '/api/memory/ingest-file') {
    sendJson(res, 404, { error: 'Not found' });
    return;
  }
  if (!engine) {
    sendJson(res, 503, { error: 'Memory not initialized' });
    return;
  }

  let body = '';
  req.on('data', (chunk: Buffer) => { body += chunk.toString(); });
  req.on('end', () => {
    try {
      const parsed = JSON.parse(body);
      const files: IngestFileRequest[] = parsed.files || [parsed];
      const results = files.map(f => ingestFile(f, engine, workspace));
      const ingested = results.filter(r => !r.skipped).length;
      const skipped = results.filter(r => r.skipped).length;
      sendJson(res, 200, { total: results.length, ingested, skipped, results });
    } catch (err: any) {
      sendJson(res, 400, { error: `Invalid JSON: ${err.message}` });
    }
  });
}

function ingestFile(req: IngestFileRequest, engine: MemoryEngine, workspace: string): IngestFileResult {
  const filePath = resolvePath(req.file_path, workspace);
  if (!filePath) {
    return { file_path: req.file_path, entries_created: 0, skipped: true, reason: 'file not found' };
  }

  const content = fs.readFileSync(filePath, 'utf-8');
  const type = req.type || 'CONTEXT';
  const tier = typeToTier(type);

  // Simple markdown section splitting + insert
  const sections = splitMarkdownSections(content);
  let created = 0;
  for (const section of sections) {
    if (section.content.trim().length === 0) continue;
    const summary = section.heading
      ? `${section.heading}: ${section.content.substring(0, 100)}`
      : section.content.substring(0, 120);
    engine.knowledge.insert({
      content: section.content,
      summary,
      type,
      tier,
      source: req.file_path,
      tags: section.heading || ''
    });
    created++;
  }

  return { file_path: req.file_path, entries_created: created, skipped: false };
}

function splitMarkdownSections(text: string): Array<{ heading: string; content: string }> {
  const lines = text.split('\n');
  const sections: Array<{ heading: string; content: string }> = [];
  let heading = '';
  let buf: string[] = [];

  for (const line of lines) {
    if (line.startsWith('#')) {
      if (buf.length > 0) sections.push({ heading, content: buf.join('\n') });
      heading = line.replace(/^#+\s*/, '');
      buf = [];
    } else {
      buf.push(line);
    }
  }
  if (buf.length > 0) sections.push({ heading, content: buf.join('\n') });
  return sections;
}

function typeToTier(type: string): string {
  switch (type) {
    case 'REQUIREMENT': case 'ARCHITECTURE': case 'PROCEDURE':
      return 'SEMANTIC';
    case 'DECISION': case 'LESSON_LEARNED':
      return 'EPISODIC';
    default:
      return 'WORKING';
  }
}

function resolvePath(filePath: string, workspace: string): string | null {
  if (path.isAbsolute(filePath) && fs.existsSync(filePath)) return filePath;
  if (workspace) {
    const wsPath = path.join(workspace, filePath);
    if (fs.existsSync(wsPath)) return wsPath;
  }
  return null;
}

function sendJson(res: http.ServerResponse, status: number, data: unknown): void {
  const body = JSON.stringify(data);
  res.writeHead(status, {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*'
  });
  res.end(body);
}
